File: cli-protonj2/src/test/java/com/redhat/mqe/MainTest.java
/*
 * Copyright (c) 2021 Red Hat, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.redhat.mqe;

import com.google.common.truth.Truth;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.internal.Intrinsics;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.*;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.function.Executable;
import util.Broker;
import util.BrokerFixture;

import java.security.Permission;
import java.util.concurrent.TimeUnit;

class SystemExitingWithStatus extends RuntimeException {
    public final int status;

    public SystemExitingWithStatus(int status) {
        this.status = status;
    }
}

class NoExitSecurityManager extends SecurityManager {
    private final SecurityManager parentManager;

    @Override
    public void checkPermission(Permission perm) {
        // allow all
    }

    @Override
    public void checkPermission(Permission perm, Object context) {
        // allow all
    }

    @Override
    public void checkExit(int status) {
        throw new SystemExitingWithStatus(status);
    }

    NoExitSecurityManager(SecurityManager parentManager) {
        this.parentManager = parentManager;
    }

    public static void assertSystemExit(int status, @NotNull Executable executable) throws Throwable {
        Intrinsics.checkNotNullParameter(executable, "executable");
        SecurityManager previousManager = System.getSecurityManager();

        try {
            NoExitSecurityManager manager = new NoExitSecurityManager(previousManager);
            System.setSecurityManager(manager);
            executable.execute();
            Assertions.fail("expected exception");
        } catch (SystemExitingWithStatus exception) {
            Truth.assertThat(exception.status).isEqualTo(status);
        } finally {
            System.setSecurityManager(previousManager);
        }
    }

    public static void assertNoSystemExit(@NotNull Function0 executable) {
        Intrinsics.checkNotNullParameter(executable, "executable");
        SecurityManager previousManager = System.getSecurityManager();

        try {
            NoExitSecurityManager manager = new NoExitSecurityManager(previousManager);
            System.setSecurityManager((SecurityManager) manager);
            executable.invoke();
        } catch (SystemExitingWithStatus var5) {
            Assertions.fail("System.exit has been called");
        } finally {
            System.setSecurityManager(previousManager);
        }
    }
}

class MainTest {
    @BeforeAll
    static void configureLogging() {
        // turn on extra verbose logging
        //Broker.configureLogging();
    }

    void checkMainInvocation(String cmd) throws Throwable {
        String[] args = cmd.split(" ");
        NoExitSecurityManager.assertSystemExit(0, () -> {
            Main.main(args);
        });
    }

    @Test
    @Timeout(value = 60, unit = TimeUnit.SECONDS)
    @ExtendWith(BrokerFixture.class)
    void testAckWithSessionClose(@BrokerFixture.TempBroker Broker broker) throws Throwable {
        broker.configuration.setSecurityEnabled(false);
        broker.configuration.setPersistenceEnabled(false); // this, or tmpdir, otherwise test runs interact
        broker.startBroker();

        // todo: have to handle amqp:// prefix
        String brokerUrl = "localhost:" + broker.addAMQPAcceptor();

        NoExitSecurityManager.assertSystemExit(0, () -> {
            Main.main(
                "sender", "--log-msgs=dict", "--broker=" + brokerUrl, "--conn-username=tckuser", "--conn-password=tckuser", "--address=test_default_username_right_password_right", "--count=1"
            );
        });

        NoExitSecurityManager.assertSystemExit(0, () -> {
            Main.main(
                "sender", "--timeout=2", "--log-msgs=dict", "--broker=" + brokerUrl, "--conn-auth-mechanisms=PLAIN", "--conn-username=admin", "--conn-password=admin", "--address=test_direct_transient_empty_message_with_string_property", "--count=10", "--msg-property=key1=value1"
            );
        });

        NoExitSecurityManager.assertSystemExit(0, () -> {
            Main.main(
                "receiver", "--timeout=2", "--log-msgs=dict", "--broker=" + brokerUrl, "--conn-auth-mechanisms=PLAIN", "--conn-username=admin", "--conn-password=admin", "--address=test_direct_transient_empty_message_with_string_property", "--count=10"
            );
        });

        NoExitSecurityManager.assertSystemExit(0, () -> {
            Main.main(
                "sender", "--log-msgs=dict", "--broker=" + brokerUrl, "--conn-auth-mechanisms=PLAIN", "--conn-username=admin", "--conn-password=admin", "--address=test_direct_transient_text_message", "--count=1", "--msg-content=Simple Text Message", "--msg-correlation-id=corr-id-JWXoIk"
            );
        });

        // test_direct_transient_text_message
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_direct_transient_text_message --count 1 --msg-content SimpleTextMessage --msg-correlation-id corr-id-eqa9vp");
        checkMainInvocation("receiver --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_direct_transient_text_message --count 1");

        // test_publish_subscribe_string
        Thread t = new Thread(() -> {
            try {
                checkMainInvocation("receiver --timeout 100 --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address topic://test_publish_subscribe_string --count 3");
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        });
        Thread.sleep(100);  // do I really want to do things like this? need better check I have a subscriber on broker
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address topic://test_publish_subscribe_string --count 3 --msg-content ABC --msg-correlation-id some-corr-id");
        t.join();
    }

    @Test
    @Disabled("These commands take way too long to execute, now that --duration works how it should")
    @Timeout(value = 60, unit = TimeUnit.SECONDS)
    @ExtendWith(BrokerFixture.class)
    void test2(@BrokerFixture.TempBroker Broker broker) throws Throwable {
        broker.configuration.setSecurityEnabled(false);
        broker.configuration.setPersistenceEnabled(false); // this, or tmpdir, otherwise test runs interact
        broker.startBroker();

        // todo: have to handle amqp:// prefix
        String brokerUrl = "localhost:" + broker.addAMQPAcceptor();


        checkMainInvocation("receiver --timeout 2 --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address JAMQMsgPatterns111Tests_test_browse_messages --recv-browse true --count 20");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_direct_transient_list_message
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_direct_transient_list_message --count 1 --msg-content-list-item  --msg-content-list-item String --msg-content-list-item ~1 --msg-content-list-item ~1.0 --msg-content-list-item 1 --msg-content-list-item 1.0 --msg-content-list-item ~-1 --msg-content-list-item ~-1.3 --msg-content-list-item -1 --msg-content-list-item ~~1 --msg-correlation-id corr-id-Mee2YQ");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_direct_transient_map_message
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_direct_transient_map_message --count 1 --msg-content-map-item empty_string= --msg-content-map-item string=String --msg-content-map-item int~1 --msg-content-map-item float~1.0 --msg-content-map-item string_int=1 --msg-content-map-item string_float=1.0 --msg-content-map-item negative_int~-1 --msg-content-map-item negative_float~-1.3 --msg-content-map-item string_negative_int=-1 --msg-content-map-item string_retype_operator=~1 --msg-correlation-id corr-id-JQt4JM");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_dead_letter_queue_with_expired_messages
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_dead_letter_queue_with_expired_messages --count 3 --msg-content ABC --msg-durable yes --msg-ttl 1000");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_message_group_simple
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_message_group_simple --count 1 --msg-content B-0 --msg-group-id B");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_reply_to_address
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_reply_to_address --count 1 --msg-reply-to test_reply_to_address-replyQ --msg-content-map-item text=replyQ_SLjFkenkBi");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_publish_subscribe_int
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address topic://test_publish_subscribe_int --count 3 --msg-content 12345 --content-type int --msg-correlation-id some-corr-id");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_publish_subscribe_map_list
        // --msg-content-map-item "key3~[123, 3.14]"

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_message_group_consumer_disconnect
        // TODO msgcontent %d feature
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_message_group_consumer_disconnect --count 40 --msg-content A-%d --duration 20 --msg-group-id A");

        // tests.JAMQMsgPatterns000Tests.JAMQMsgPatternsTests.test_reply_to_address
        checkMainInvocation("receiver --timeout 5 --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_reply_to_address --count 1 --process-reply-to");

        // tests.JAMQNode000Tests.JAMQNodeTests.test_address_full_policy_block
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address JAMQNode111Tests_test_address_full_policy_block --count 1 --msg-content-from-file /etc/passwd");

        // tests.JAMQNode000Tests.JAMQNodeTests.test_direct_transient_large_string_message_size_1mb
        // ditto for receiver
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_direct_transient_large_string_message_size_1mb --count 1 --msg-content-from-file /etc/passwd --msg-content-hashed True");

        // tests.JAMQNode000Tests.JAMQNodeTests.test_max_consumers_queue
        // TODO: client is not started at all in the test; some issues with params mapping?

        // tests.JAMQMessage000Tests.JAMQMessageTests.test_amqp_bare_message_consistency
        // Unknown options: '--msg-subject', 'amqp_bare_message_test', '--msg-user-id', 'admin', '--msg-priority', '7', '--conn-populate-user-id', 'True', '--msg-group-seq', '1', '--msg-reply-to-group-id', 'group-a'
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_amqp_bare_message_consistency --count 1 --msg-subject amqp_bare_message_test --msg-reply-to ExpiryQueue --msg-property PI=~3.141592 --msg-property color=red --msg-property mapKey=mapValue --msg-content amqp_bare_msg-CBJJIY --msg-durable True --msg-ttl 300000 --msg-correlation-id amqp_bare_msg-CBJJIY --msg-user-id admin --msg-priority 7 --conn-populate-user-id True --msg-group-id group-a --msg-group-seq 1 --msg-reply-to-group-id group-a");

        // tests.JAMQMessage000Tests.JAMQMessageTests.test_populate_validated_user_option
        // TODO FAIL     dtestlib.Test:levels.py:61 Checking properties keys for validated user with expected '_AMQ_VALIDATED_USER' or 'JMSXUserID': dict_keys([]) # result:False (exp. True), dur.:-1.00 err_cnt:1

        // tests.JAMQMessage000Tests.JAMQMessageTests.test_scheduled_message_zero_timestamp
        // Unknown options: '--msg-id'
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address JAMQMessage111Tests_test_scheduled_message_zero_timestamp --count 2 --msg-id ReferenceMessage_JAMQMessage111Tests_test_scheduled_message_zero_timestamp --msg-property _AMQ_SCHED_DELIVERY~0");

        // tests.JAMQMessage000Tests.JAMQMessageTests.test_client_acknowledge_inactivity_exception
        // --ssn-ack-mode
        // --duration --duration-mode --ssn-ack-mode
        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_client_acknowledge_inactivity_exception --count 20 --msg-durable True --ssn-ack-mode client");
        checkMainInvocation("receiver --timeout 10 --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_client_acknowledge_inactivity_exception --count 20 --duration 100 --duration-mode after-receive --ssn-ack-mode client");


//
//        checkMainInvocation("sender --log-msgs dict --broker " + brokerUrl + " --conn-auth-mechanisms PLAIN --conn-username admin --conn-password admin --address test_direct_transient_text_message --count 1 --msg-content SimpleTextMessage --msg-correlation-id corr-id-eqa9vp");
    }

//    void testMessageContentListItem() {
//        '--msg-content-list-item', '', '--msg-content-list-item', 'String', '--msg-content-list-item', '~1', '--msg-content-list-item', '~1.0', '--msg-content-list-item', '1', '--msg-content-list-item', '1.0', '--msg-content-list-item', '~-1', '--msg-content-list-item', '~-1.3', '--msg-content-list-item', '-1', '--msg-content-list-item', '~~1'
//    }
}

File: cli-qpid-jms/src/test/kotlin/MainTest.kt
/*
 * Copyright (c) 2017 Red Hat, Inc.
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import com.redhat.mqe.ClientListener
import com.redhat.mqe.jms.Main
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertTimeoutPreemptively
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.function.Executable
import java.time.Duration

class AacMainTest : AbstractMainTest() {

    override val brokerUrl = "amqp://127.0.0.1:5672"
    override val sslBrokerUrl = "amqps://127.0.0.1:5673"

    override val senderAdditionalOptions = """
--conn-tcp-sock-linger 1000
--capacity 1
--conn-async-send true
--conn-auth-mechanisms anonymous
--conn-auth-sasl false
--conn-clientid aClientId
--conn-clientid-prefix aClientIdPrefix
--conn-close-timeout 1000
--conn-conn-timeout 1000
--conn-connid-prefix aConnIdPrefix
--conn-heartbeat 1000
--conn-local-msg-priority true
--conn-max-frame-size 4096
--conn-prefetch 1
--conn-prefetch-browser 1
--conn-prefetch-queue 1
--conn-prefetch-topic 1
--conn-prefetch-topic-dur 1
--conn-queue-prefix aQueuePrefix
--conn-reconnect true
--conn-reconnect-backoff false
--conn-reconnect-backoff-multiplier 1
--conn-reconnect-initial-delay 1
--conn-reconnect-interval 1000
--conn-reconnect-limit 1000
--conn-reconnect-start-limit 1000
--conn-reconnect-timeout 1000
--conn-reconnect-warn-attempts 1
--conn-redeliveries-max 1
--conn-tcp-buf-size-recv 1
--conn-tcp-buf-size-send 1
--conn-tcp-conn-timeout 1000
--conn-tcp-keep-alive true
--conn-tcp-no-delay false
--conn-tcp-sock-timeout 1000
--conn-tcp-traffic-class 1
--conn-topic-prefix aTopicPrefix
--conn-valid-prop-names false
--msg-content-type aMsgContentType
--msg-correlation-id aCorrelationId
--msg-durable false
--msg-group-id aMsgGroupId
--msg-group-seq -1
--msg-no-timestamp true
--msg-priority 1
--timeout 2
--tx-size 1
--msg-reply-to aReplyToQueue
--msg-reply-to-group-id aReplyToGroupId
--msg-subject aMsgSubject
--msg-ttl 10000
--msg-user-id aMsgUserId
--property-type String
""".split(" ", "\n").toTypedArray()


// cannot set Client ID, because more than one connection is created, and these would clash
//--conn-clientid aClientId
    override val connectorAdditionalOptions = """
--conn-async-send true
--conn-auth-mechanisms anonymous
--conn-auth-sasl false
--conn-clientid-prefix aClientIdPrefix
--conn-close-timeout 1000
--conn-conn-timeout 1000
--conn-connid-prefix aConnIdPrefix
--conn-heartbeat 1000
--conn-local-msg-priority true
--conn-max-frame-size 4096
--conn-prefetch 1
--conn-prefetch-browser 1
--conn-prefetch-queue 1
--conn-prefetch-topic 1
--conn-prefetch-topic-dur 1
--conn-queue-prefix aQueuePrefix
--conn-reconnect true
--conn-reconnect-backoff false
--conn-reconnect-backoff-multiplier 1
--conn-reconnect-initial-delay 1
--conn-reconnect-interval 1000
--conn-reconnect-limit 1000
--conn-reconnect-start-limit 1000
--conn-reconnect-timeout 1000
--conn-reconnect-warn-attempts 1
--conn-redeliveries-max 1
--conn-tcp-buf-size-recv 1
--conn-tcp-buf-size-send 1
--conn-tcp-conn-timeout 1000
--conn-tcp-keep-alive true
--conn-tcp-no-delay false
--conn-tcp-sock-timeout 1000
--conn-tcp-traffic-class 1
--conn-topic-prefix aTopicPrefix
--conn-valid-prop-names false
""".split(" ", "\n").toTypedArray()

    override fun main_(listener: ClientListener, args: Array<String>) = Main.main(listener, args)

    /**
     * transport.logBytes option in qpid-jms
     */
    @Test
    fun sendAndReceiveSingleMessageLogBytes() {
        val senderParameters =
            "sender --log-msgs dict --out json --broker $brokerUrl --address $address --count 1 --log-bytes".split(" ").toTypedArray()
        val receiverParameters =
            "receiver --log-msgs dict --out json --broker $brokerUrl --address $address --count 1 --log-bytes".split(" ").toTypedArray()
        Assertions.assertTimeoutPreemptively(Duration.ofSeconds(10)) {
            print("Sending: ")
            main(senderParameters)
            print("Receiving: ")
            main(receiverParameters)
        }
    }

    // https://issues.apache.org/jira/browse/QPIDJMS-342
    @Tag("issue")
    @Test
    fun attemptConnectingToWrongPortWithTimeout() {
        val senderParameters =
            ("connector" +
                " --broker-uri failover:(amqp://127.0.0.1:1883)?jms.connectTimeout=1&failover.maxReconnectAttempts=1&failover.startupMaxReconnectAttempts=1" +
                ""
                ).split(" ").toTypedArray()
        assertTimeoutPreemptively(Duration.ofSeconds(5)) {
            assertSystemExit(1, Executable {
                main(senderParameters)
            })
        }
    }
}

//        @Test fun sendAndReceiveSingleMessageSsl() {
//        val senderParameters =
//                "sender --log-msgs dict --broker amqps://127.0.0.1:61616 --address lalaLand --count 1 --conn-ssl-truststore-location /home/jdanek/Downloads/AMQ7/amq7cr1i0/redhatqe.truststore --conn-ssl-truststore-password password --conn-ssl-keystore-location /home/jdanek/Downloads/AMQ7/amq7cr1i0/client.keystore --conn-ssl-keystore-password password".split(" ").toTypedArray()
//        val receiverParameters =
//                "receiver --log-msgs dict --broker amqp://127.0.0.1:5672 --address lalaLand --count 1".split(" ").toTypedArray()
//        print("Sending: ")
//        Main.main(senderParameters)
////        print("Receiving: ")
////        Main.main(receiverParameters)
//    }
//
//    @Test fun reconnect() {
//        val host = "amqp://127.0.0.1:61616"
//        val address = "reconnect"
//        Main.main(
//                "receiver  --timeout 120 --log-msgs dict --broker $host --conn-reconnect True --conn-username admin --conn-password admin --address $address --count 1".split(" ").toTypedArray()
//        )
//    }
//
//    @Test fun isAddressMulticast() {
////        val host = "amqp://172.17.0.4:5672 --conn-username a --conn-password b"
//        val host = "amqp://172.17.0.4:5672"
//        val address = "multicast"
//
//        val r1 = Thread {
//            Main.main(
//                    "receiver --log-msgs dict --broker $host --address $address --count 2 --timeout 10".split(" ").toTypedArray()
//            )
//        }
//
//        val r2 = Thread {
//            Main.main(
//                    "receiver --log-msgs dict --broker $host --address $address --count 2 --timeout 10".split(" ").toTypedArray()
//            )
//
//        }
//
//        val s1 = Thread {
//            Main.main(
//                    "sender --log-msgs dict --broker $host --address $address --count 2 --msg-content AMQP%d".split(" ").toTypedArray()
//            )
//        }
//
//        r1.start()
//        r2.start()
//        Thread.sleep(1 * 1000)
//        s1.start()
//
//        s1.join()
//        r1.join()
//        r2.join()
//    }
//
//    @Test fun sendSingleMessageToTopicAndRollback() {
//        val senderParameters =
//                ("sender --log-msgs dict --broker amqp://127.0.0.1:5672 --address topic://topic --count 1 --tx-action rollback tx-endloop-action rollback".split(" ").toTypedArray())
//        print("Sending: ")
//        Main.main(senderParameters)
//    }
//}
